File: AUGUST-2025/03-08-2025/p2106.rb
class Numeric
    def max(v) = (self > v ? self : v)
    def min(v) = (self < v ? self : v)
end

class Solver
    attr_reader :fruits, :start_pos

    def initialize(fruits, start_pos)
        @fruits, @start_pos = fruits, start_pos
    end

    def min_s(lpos, rpos)
        d = rpos - lpos
        ((start_pos - lpos).abs + d).min((start_pos - rpos).abs + d)
    end

    def solve(k)
        l, sum = 0, 0
        fruits.each_with_index.inject(0) {|res, ((pos, f), i)|
            sum += f
            sum -= fruits[(l += 1) - 1].last while l <= i && min_s(fruits[l].first, pos) > k
            res.max(sum)
        }
    end
end

def max_total_fruits(fruits, start_pos, k) = Solver.new(fruits, start_pos).solve(k)

File: AUGUST-2025/03-08-2025/p2106.py
class Solution:
    def maxTotalFruits(self, fruits: list[list[int]], startPos: int, k: int) -> int:

        def steps(lpos: int, rpos: int) -> int:
            if rpos <= startPos:
                return startPos - lpos
            elif lpos >= startPos:
                return rpos - startPos
            else:
                return min(startPos + rpos - 2 * lpos,
                           2 * rpos - startPos - lpos)

        cur_sum = 0
        ans = 0
        l = 0
        n = len(fruits)

        for r in range(n):
            cur_sum += fruits[r][1]

            while l <= r and steps(fruits[l][0], fruits[r][0]) > k:
                cur_sum -= fruits[l][1]
                l += 1

            ans = max(ans, cur_sum)

        return ans